{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "877ec869-c7b2-4cef-b434-3d2d95baa674",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: SeuratObject\n",
      "\n",
      "Loading required package: sp\n",
      "\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, t\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘cowplot’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:ggpubr’:\n",
      "\n",
      "    get_legend\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘rstatix’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    filter\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(Seurat)\n",
    "library(liana)\n",
    "library(dplyr)\n",
    "library(tidyr)\n",
    "library(ggplot2)\n",
    "library(ggpubr)\n",
    "library(cowplot)\n",
    "library(rstatix)\n",
    "\n",
    "`%notin%` <- Negate(`%in%`)\n",
    "set.seed(123)\n",
    "\n",
    "root.dir <- \"/project/bicistronic_carT_gbm_Jackie/\"\n",
    "figures.dir <- paste0(root.dir, \"Figures/\") # figures will be outputted in this folder\n",
    "csf.liana.output.dir <- paste0(root.dir, \"ProcessedData/\", \"liana/\", \"csf/\") # liana results (table) will be outputted to this folder "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4d1e82fa-c787-4e18-941d-1f6d0947022a",
   "metadata": {},
   "outputs": [],
   "source": [
    "all.high.dose.patients <- c(\"P1\", \"P2\", \"P3\", \"P4\", \"P5\", \"P6\", \"P7\", \"P8\", \"P9\", \"P10\", \"P11\", \"P12\")\n",
    "discovery.cohort.patients <- c(\"P6\", \"P1\", \"P8\", \"P5\", \"P2\", \"P7\", \"P4\", \"P3\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "174d8f29-05dc-477d-a92a-d7ba8ce21bdc",
   "metadata": {},
   "source": [
    "# Start here with CSF Seurat object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "64343706-50ca-4a8c-8e29-d26903407238",
   "metadata": {},
   "outputs": [],
   "source": [
    "csf <- readRDS(paste0(root.dir, \"ProcessedData/seurat/all_csf_annotated_final.rds\"))\n",
    "csf$response.status <- sapply(csf$Responder_Status, function(x) ifelse(x, \"Responder\", \"Non-responder\"))\n",
    "csf$Cohort <- sapply(csf$Patient, function(x) ifelse(x %in% discovery.cohort.patients, \"Discovery\", \"Validation\"))\n",
    "csf$Day <- factor(csf$Day, levels = c(\"D0\", \"D7\", \"D21\"))\n",
    "csf$cell_type_level3 <- sapply(csf$cell_type_level3, function(x) ifelse(x == \"CCD56bright_CD16neg_NK\", \"CD56bright_CD16neg_NK\", x)) # fix typo\n",
    "csf$cell_type <- csf$cell_type_level3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "12622bba-da69-416c-9a3f-5fcea9991045",
   "metadata": {},
   "outputs": [],
   "source": [
    "# combine monocyte/macrophage cell types together\n",
    "csf$cell_type_broad <- apply(csf@meta.data, 1, function(row) \n",
    "    case_when(\n",
    "        row[['cell_type']] %in% c('Monocyte', 'BAM', 'Microglia', 'Macrophage') ~ 'Mono/Mac',\n",
    "        TRUE ~ row[[\"cell_type\"]] \n",
    "    )\n",
    ")\n",
    "csf$cell_type_fine <- csf$cell_type\n",
    "csf$cell_type <- csf$cell_type_broad\n",
    "cell.type.col <- \"cell_type_broad\"\n",
    "Idents(csf) <- cell.type.col"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "91edb5c8-3a25-470e-92b0-53011056cd18",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a map of patient to response status\n",
    "patient.response <- unique(paste0(csf$Patient, \"_\", csf$Responder_Status))\n",
    "split_data <- strsplit(patient.response, \"_\")\n",
    "patient_response <- setNames(\n",
    "  sapply(split_data, function(x) ifelse(x[2] == \"TRUE\", \"Responder\", \"Non-responder\")),\n",
    "  sapply(split_data, function(x) x[1])\n",
    ")\n",
    "patient_map <- tibble(\n",
    "  Patient = names(patient_response),\n",
    "  response.status = unlist(patient_response)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "0030230c-2f66-4401-a37f-fce8606029a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# subset to high dose only\n",
    "sample_ids <- unique(csf$sample_id)\n",
    "csf.high.dose <- subset(csf, subset = Dose_Class == \"High\")\n",
    "sample_ids_high_dose <- unique(csf.high.dose$sample_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "51428570-f359-4e10-a3b3-39fcdc8ab1fb",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"P1D0\"\n",
      "[1] \"P1D7\"\n",
      "[1] \"P1D21\"\n",
      "[1] \"P2D0\"\n",
      "[1] \"P2D7\"\n",
      "[1] \"P2D21\"\n",
      "[1] \"P3D0\"\n",
      "[1] \"P3D7\"\n",
      "[1] \"P3D21\"\n",
      "[1] \"P4D0\"\n",
      "[1] \"P4D7\"\n",
      "[1] \"P4D21\"\n",
      "[1] \"P5D0\"\n",
      "[1] \"P5D7\"\n",
      "[1] \"P5D21\"\n",
      "[1] \"P6D0\"\n",
      "[1] \"P6D7\"\n",
      "[1] \"P6D21\"\n",
      "[1] \"P7D0\"\n",
      "[1] \"P7D7\"\n",
      "[1] \"P7D21\"\n",
      "[1] \"P8D0\"\n",
      "[1] \"P8D7\"\n",
      "[1] \"P8D21\"\n",
      "[1] \"P9D0\"\n",
      "[1] \"P9D7\"\n",
      "[1] \"P10D0\"\n",
      "[1] \"P10D7\"\n",
      "[1] \"P11D0\"\n",
      "[1] \"P11D7\"\n",
      "[1] \"P12D0\"\n",
      "[1] \"P12D7\"\n",
      "[1] \"P13D0\"\n",
      "[1] \"P13D7\"\n",
      "[1] \"P14D0\"\n",
      "[1] \"P14D7\"\n",
      "[1] \"P15D0\"\n",
      "[1] \"P15D7\"\n",
      "[1] \"P16D0\"\n",
      "[1] \"P16D7\"\n",
      "[1] \"P17D0\"\n",
      "[1] \"P17D7\"\n",
      "[1] \"P18D0\"\n",
      "[1] \"P18D7\"\n"
     ]
    }
   ],
   "source": [
    "# run liana on each sample\n",
    "for (sample_id_use in sample_ids) {\n",
    "    print(sample_id_use)\n",
    "    \n",
    "    filename <- paste0(csf.liana.output.dir, sample_id_use, \"_liana_test.tsv\")\n",
    "\n",
    "    # read in file if it already exists\n",
    "    if (file.exists(filename)) {\n",
    "        next\n",
    "    }\n",
    "\n",
    "    csf_subset = subset(csf, subset = sample_id == sample_id_use)\n",
    "    print(dim(csf_subset))\n",
    "    print(table(csf_subset$cell_type_broad))\n",
    "    \n",
    "    liana_test <- liana_wrap(csf_subset, idents_col = cell.type.col)\n",
    "    liana_test <- liana_test %>%\n",
    "        liana_aggregate()\n",
    "    write.table(liana_test, filename, sep = \"\\t\", quote=F, col.names=T, row.names=F)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "21d654d5-f6d3-4e71-91c6-ddfb12f76ccb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in LIANA results and aggregate all samples\n",
    "filename <- paste0(csf.liana.output.dir, \"all_samples_combined_unfiltered_liana.tsv\")\n",
    "if (!file.exists(filename)) {\n",
    "    \n",
    "    liana_combined <- data.frame()\n",
    "    for (sample_id_use in sample_ids_high_dose) {\n",
    "    \n",
    "        liana_test <- read.csv(paste0(csf.liana.output.dir, sample_id_use, \"_liana_test.tsv\"), sep = \"\\t\")\n",
    "    \n",
    "        patient_id <- unlist(str_split(sample_id_use, \"D\"))[1]\n",
    "        day <- paste0(\"D\", unlist(str_split(sample_id_use, \"D\"))[2])\n",
    "    \n",
    "        # add metadata\n",
    "        liana_test$source <- sapply(liana_test$source, function(x) ifelse(x == \"CCD56bright_CD16neg_NK\", \"CD56bright_CD16neg_NK\", x)) # fix typo\n",
    "        liana_test$target <- sapply(liana_test$target, function(x) ifelse(x == \"CCD56bright_CD16neg_NK\", \"CD56bright_CD16neg_NK\", x)) # fix typo\n",
    "                                    \n",
    "        liana_test$sample_id <- sample_id_use\n",
    "        liana_test$Patient <- patient_id\n",
    "        liana_test$Day <- day\n",
    "\n",
    "        # add reponse status\n",
    "        response.status <- patient_map %>%\n",
    "          dplyr::filter(Patient == patient_id) %>%\n",
    "          dplyr::pull(response.status)\n",
    "        liana_test$response.status <- response.status\n",
    "        \n",
    "        liana_combined <- rbind(liana_combined, liana_test)\n",
    "    }\n",
    "    liana_combined <- liana_combined %>% mutate(label = paste0(source, \":\", ligand.complex, \" -> \", target, \":\", receptor.complex))\n",
    "    liana_combined <- liana_combined %>% mutate(interaction.label = paste0(ligand.complex, \" -> \", receptor.complex))\n",
    "    liana_combined$significance.level <- sapply(liana_combined$aggregate_rank, function(x) case_when(\n",
    "        x <= 0.05 ~ \"<= 0.05\",\n",
    "        x <= 0.1 ~ \"<= 0.1\",\n",
    "        TRUE ~ \"> 0.1\"\n",
    "    ))\n",
    "    \n",
    "    write.table(liana_combined, filename, sep = \"\\t\", quote=F, col.names=T, row.names=F)\n",
    "} else {\n",
    "\n",
    "    # read in aggregated file if it already exists\n",
    "    liana_combined <- read.csv(filename, sep = \"\\t\")\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "478c261a-07df-4767-b93b-37587629325e",
   "metadata": {},
   "outputs": [],
   "source": [
    "rm(csf) # not used downstream\n",
    "all.patients <- all.high.dose.patients"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ff1ac317-bc8d-404c-a446-2dd4f871b832",
   "metadata": {},
   "outputs": [],
   "source": [
    "liana_combined$Day <- factor(liana_combined$Day, levels=c(\"D0\", \"D7\", \"D21\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3ee9d81e-878d-4513-8f85-4fcfb3989b8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# swap CTLA4 from receptor to ligand\n",
    "swapped.df <- liana_combined %>% filter(receptor.complex == \"CTLA4\") %>% rename(target = source, source = target, receptor.complex = ligand.complex, ligand.complex = receptor.complex)\n",
    "other.df <- liana_combined %>% filter(receptor.complex != \"CTLA4\")\n",
    "liana_combined <- rbind(other.df, swapped.df)\n",
    "liana_combined <- liana_combined %>% mutate(label = paste0(source, \":\", ligand.complex, \" -> \", target, \":\", receptor.complex))\n",
    "liana_combined <- liana_combined %>% mutate(interaction.label = paste0(ligand.complex, \" -> \", receptor.complex))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "78277c1e-f31b-4968-b961-7d5b579bed13",
   "metadata": {},
   "outputs": [],
   "source": [
    "immune.checkpoints <- list(c(\"TGFB1\", \"TGFBR1_TGFBR2\"),\n",
    "                           c(\"BTLA\", \"TNFRSF14\"),\n",
    "                           c(\"TNFSF18\", \"TNFRSF18\"),\n",
    "                           c(\"CTLA4\", \"CD80\"),\n",
    "                           c(\"CTLA4\", \"CD86\"),\n",
    "                           c(\"ENTPD1\", \"ADORA2A\"),\n",
    "                           c(\"CD274\", \"PDCD1\"),\n",
    "                           c(\"PDCD1LG2\", \"PDCD1\"),\n",
    "                           c(\"LGALS9\", \"HAVCR2\"),\n",
    "                           c(\"HLA-DPA1\", \"LAG3\"),\n",
    "                           c(\"CD80\", \"CD28\"),\n",
    "                           c(\"CD86\", \"CD28\"),\n",
    "                           c(\"CD40LG\", \"CD40\"),\n",
    "                           c(\"TNFSF4\", \"TNFRSF4\"), # OX40\n",
    "                           c(\"TNFSF9\", \"TNFRSF9\"), # 41BB\n",
    "                           c(\"ICOSLG\", \"CD278\"), # ICOS\n",
    "                           c(\"CD70\", \"CD27\")\n",
    "                          )\n",
    "canonical.interactions <- sapply(immune.checkpoints, function(x) paste0(x[1], \" -> \", x[2])) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "813261d2-b558-4793-b5f4-0f7b8420b9a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot.canonical.interactions <- function(liana_df, filepath, cell.types, interactions.use=immune.checkpoints, \n",
    "                                        levels.interactions=canonical.interactions, \n",
    "                                        levels.celltypes = NA,\n",
    "                                        facet.by.source=T, timepoint.var = \"Day\",\n",
    "                                        plot.width = 30, plot.height = NA) {\n",
    "\n",
    "    \n",
    "    plot.list <- list()\n",
    "    for (s in cell.types) {\n",
    "\n",
    "        if (facet.by.source) {\n",
    "            liana_filter_df <- liana_df %>% dplyr::filter(source==s)\n",
    "        } else {\n",
    "            liana_filter_df <- liana_df %>% dplyr::filter(target==s)\n",
    "        }\n",
    "        \n",
    "        interactions.df <- data.frame()\n",
    "        for (i in 1:length(interactions.use)) {\n",
    "            ligand <- interactions.use[[i]][1]\n",
    "            receptor <- interactions.use[[i]][2]\n",
    "            temp <- liana_filter_df %>% dplyr::filter(ligand.complex == ligand, receptor.complex == receptor)\n",
    "            interactions.df <- rbind(interactions.df, temp)\n",
    "        }\n",
    "        # add 0 for missing data\n",
    "        interactions.df <- interactions.df %>% complete(source, target, nesting(ligand.complex, receptor.complex, interaction.label), \n",
    "                                                        .data[[timepoint.var]], nesting(Patient, response.status), fill = \n",
    "                                                        list(sca.LRscore = 0, natmi.edge_specificity = 0))\n",
    "\n",
    "\n",
    "        interactions.df$y <- interactions.df$sca.LRscore \n",
    "    \n",
    "        df <- interactions.df \n",
    "\n",
    "        if(nrow(interactions.df) == 0) {\n",
    "            next\n",
    "        }\n",
    "        \n",
    "        if (!all(is.na(levels.interactions))) {\n",
    "            df$interaction.label <- factor(df$interaction.label, levels = canonical.interactions)\n",
    "        }\n",
    "\n",
    "        if (!all(is.na(levels.celltypes))) {\n",
    "            df$target <- factor(df$target, levels = levels.celltypes)\n",
    "        }\n",
    "    \n",
    "        df$`Interaction strength` <- df$sca.LRscore \n",
    "        df$`Interaction specificity` <- df$natmi.edge_specificity\n",
    "        df$Patient <- factor(df$Patient, levels = all.patients)\n",
    "\n",
    "        if (facet.by.source) {\n",
    "            x <- \"target\"\n",
    "            x.label <- \"Target\"\n",
    "        } else {\n",
    "            x <- \"source\"\n",
    "            x.label <- \"Source\"\n",
    "        }\n",
    "\n",
    "        p <- ggplot(df, aes(.data[[x]], interaction.label)) +\n",
    "            geom_point(aes(\n",
    "                           fill = `Interaction strength`, \n",
    "                           size = `Interaction specificity`),\n",
    "                      shape = 21, stroke = 0.8) +\n",
    "            theme_bw() +\n",
    "            theme(axis.text.x = element_text(angle = 45, hjust = 1)) +\n",
    "            scale_fill_viridis_c(limits = c(0, 1)) +\n",
    "            scale_size(\n",
    "                limits = c(0, 1),\n",
    "                range = c(0, 10),    \n",
    "                breaks = c(0, 0.2, 0.4, 0.6, 0.8, 1),\n",
    "            ) +\n",
    "            facet_grid(cols=vars(Patient), rows=vars(.data[[timepoint.var]])) +\n",
    "            xlab(x.label) +\n",
    "            ylab(\"Interaction\")\n",
    "\n",
    "        plot.list[[s]] <- p\n",
    "    }\n",
    "\n",
    "    if (length(plot.list) == 0) {\n",
    "        print(\"no interactions\")\n",
    "    } else {\n",
    "        if (length(plot.list) > 1) {                                                              \n",
    "            p <- plot_grid(plotlist = plot.list, ncol = 1, labels = names(plot.list))\n",
    "        } else {\n",
    "            p <- p + ggtitle(names(plot.list))\n",
    "        }\n",
    "        \n",
    "        if (is.na(plot.height)) {\n",
    "            plot.height <- length(plot.list) * 10\n",
    "        }\n",
    "        pdf(filepath, width = plot.width, height = plot.height)\n",
    "        print(p)\n",
    "        dev.off()\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "2468c81e-8025-4f6b-be55-6e3a339d37d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot.canonical.interactions.patients.aggregated <- function(liana_df, filepath, cell.types, timepoint.of.interest, interactions.use=immune.checkpoints, \n",
    "                                        levels.interactions=canonical.interactions, \n",
    "                                        levels.celltypes = NA,\n",
    "                                        facet.by.source=T, timepoint.var = \"Day\",\n",
    "                                        plot.width = 30, plot.height = NA) {\n",
    "\n",
    "    plot.list <- list()\n",
    "    for (s in cell.types) {\n",
    "\n",
    "        if (facet.by.source) {\n",
    "            liana_filter_df <- liana_df %>% dplyr::filter(source==s)\n",
    "        } else {\n",
    "            liana_filter_df <- liana_df %>% dplyr::filter(target==s)\n",
    "        }\n",
    "        \n",
    "        interactions.df <- data.frame()\n",
    "        for (i in 1:length(interactions.use)) {\n",
    "            ligand <- interactions.use[[i]][1]\n",
    "            receptor <- interactions.use[[i]][2]\n",
    "            temp <- liana_filter_df %>% dplyr::filter(ligand.complex == ligand, receptor.complex == receptor, .data[[timepoint.var]] == timepoint.of.interest)\n",
    "            interactions.df <- rbind(interactions.df, temp)\n",
    "        }\n",
    "        # add 0 for missing data\n",
    "        interactions.df <- interactions.df %>% complete(source, target, nesting(ligand.complex, receptor.complex, interaction.label), \n",
    "                                                        nesting(Patient, response.status), fill = \n",
    "                                                        list(sca.LRscore = 0, natmi.edge_specificity = 0))\n",
    "\n",
    "        if(nrow(interactions.df) == 0) {\n",
    "            next\n",
    "        }\n",
    "\n",
    "        df <- interactions.df %>% \n",
    "            group_by(source, target, ligand.complex, receptor.complex, interaction.label) %>% \n",
    "            summarize(mean.sca.LRscore = mean(sca.LRscore),\n",
    "                      mean.natmi.edge_specificity = mean(natmi.edge_specificity)) %>%\n",
    "            ungroup()\n",
    "        \n",
    "        df$y <- df$mean.sca.LRscore\n",
    "                \n",
    "        if (!all(is.na(levels.interactions))) {\n",
    "            df$interaction.label <- factor(df$interaction.label, levels = canonical.interactions)\n",
    "        }\n",
    "\n",
    "        if (!all(is.na(levels.celltypes))) {\n",
    "            df$target <- factor(df$target, levels = levels.celltypes)\n",
    "        }\n",
    "        \n",
    "        df$`Interaction strength` <- df$mean.sca.LRscore \n",
    "        df$`Interaction specificity` <- df$mean.natmi.edge_specificity\n",
    "\n",
    "        if (facet.by.source) {\n",
    "            x <- \"target\"\n",
    "            x.label <- \"Target\"\n",
    "        } else {\n",
    "            x <- \"source\"\n",
    "            x.label <- \"Source\"\n",
    "        }\n",
    "\n",
    "        p <- ggplot(df, aes(.data[[x]], interaction.label)) +\n",
    "            geom_point(aes(\n",
    "                           fill = `Interaction strength`, \n",
    "                           size = `Interaction specificity`),\n",
    "                      shape = 21, stroke = 0.8) +\n",
    "            theme_bw() +\n",
    "            theme(axis.text.x = element_text(angle = 45, hjust = 1)) +\n",
    "            scale_fill_viridis_c(limits = c(0, 1)) +\n",
    "            scale_size(\n",
    "                limits = c(0, 0.1),\n",
    "                range = c(0, 8),    \n",
    "                breaks = c(0, 0.025, 0.05, 0.075, 0.1), # the size scaling can be customized as needed\n",
    "            ) +\n",
    "            ylab(\"Interaction\") +\n",
    "            xlab(x.label)\n",
    "    \n",
    "        plot.list[[s]] <- p\n",
    "    }\n",
    "    \n",
    "    if (length(plot.list) == 0) {\n",
    "        print(\"no interactions\")\n",
    "    } else {\n",
    "        if (length(plot.list) > 1) {                                                              \n",
    "            p <- plot_grid(plotlist = plot.list, ncol = 1, labels = names(plot.list))\n",
    "        } else {\n",
    "            p <- p + ggtitle(paste0(names(plot.list), \" interactions at \", timepoint.of.interest))\n",
    "        }\n",
    "        \n",
    "        if (is.na(plot.height)) {\n",
    "            plot.height <- length(plot.list) * 10\n",
    "        }\n",
    "        pdf(filepath, width = plot.width, height = plot.height)\n",
    "        print(p)\n",
    "        dev.off()\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "f29890be-ffb9-4860-b0dc-ee476032cbae",
   "metadata": {},
   "outputs": [],
   "source": [
    "# canonical Treg interactions\n",
    "treg.interactions <- list(\n",
    "                           c(\"CTLA4\", \"CD80\"),\n",
    "                           c(\"CTLA4\", \"CD86\"),\n",
    "                           c(\"LGALS9\", \"HAVCR2\"),\n",
    "                           c(\"CD70\", \"CD27\")\n",
    "                          )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "83fda520-d012-4142-9272-461487337fc3",
   "metadata": {},
   "outputs": [],
   "source": [
    "all.celltypes <- sort(unique(liana_combined$target))\n",
    "celltypes.use <- setdiff(all.celltypes, c(\"Mixed CAR+/CAR- Tfh\", \"NKT\", \"Plasma\", \"gdT\", \"ILC\", \"MAIT\", \"Choroid_Plexus_Epithelium\"))\n",
    "\n",
    "liana_combined_filtered <- liana_combined %>% filter(target %in% celltypes.use)\n",
    "liana_combined_filtered <- liana_combined_filtered %>% filter(Patient %in% all.high.dose.patients)\n",
    "liana_combined_filtered$Patient <- factor(liana_combined_filtered$Patient, levels = all.high.dose.patients)\n",
    "liana_combined_filtered$Cohort <- sapply(liana_combined_filtered$Patient, function(x) ifelse(x %in% discovery.cohort.patients, \"Discovery\", \"Validation\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "da9f07f1-c978-4fd1-921b-6acd254e6967",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Supplemental Figure 7E - Discovery cohort\n",
    "plot.canonical.interactions(liana_df=liana_combined_filtered %>% filter(Patient %in% discovery.cohort.patients, target != \"Treg\"), \n",
    "                            filepath = paste0(figures.dir, \"Supp 7E immune checkpoints Treg as source discovery.pdf\"), cell.types=c(\"Treg\"), plot.width = 40)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "8e89e1b9-ff7f-41d7-aed0-f8efc151a3f2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Supplemental Figure 7E - Validation cohort\n",
    "liana_combined_filtered_validation <- liana_combined_filtered %>% filter(Patient %notin% discovery.cohort.patients, target != \"Treg\")\n",
    "liana_combined_filtered_validation$Day <- factor(liana_combined_filtered_validation$Day, levels = c(\"D0\", \"D7\"))\n",
    "plot.canonical.interactions(liana_df=liana_combined_filtered_validation, filepath = paste0(figures.dir, \"Supp 7E immune checkpoints Treg as source validation.pdf\"), cell.types=c(\"Treg\"), plot.width = 20)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "4a97bdf5-d1c8-4c27-ac0f-5ac86ca929c1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'source', 'target', 'ligand.complex',\n",
      "'receptor.complex'. You can override using the `.groups` argument.\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 1 row containing missing values or values outside the scale range\n",
      "(`geom_point()`).”\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Figure 6G\n",
    "plot.canonical.interactions.patients.aggregated(liana_df=liana_combined_filtered %>% filter (target != \"Treg\"), filepath = paste0(figures.dir, \"6G Day 7 Treg interactions Treg as source aggregated.pdf\"), \n",
    "                            cell.types=c(\"Treg\"), timepoint.of.interest=\"D7\", interactions.use = treg.interactions, levels.celltypes = celltypes.use, plot.width=10, plot.height = 5)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "3aecdaf4-333a-4cfb-9a73-9fcf93d16db2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Figure 6H\n",
    "csf.high.dose.subset <- subset(csf.high.dose, subset = cell_type %notin% c(\"Mixed CAR+/CAR- Tfh\", \"NKT\", \"Plasma\", \"gdT\", \"ILC\", \"MAIT\", \"Choroid_Plexus_Epithelium\"))\n",
    "p <- DotPlot(csf.high.dose.subset, features = c(\"CTLA4\"), group.by = \"cell_type\") + ggtitle(\"CTLA4 expression in CSF\") + xlab(\"\") + ylab(\"\") + coord_flip() +\n",
    "  theme(axis.text.x = element_text(angle = 45, hjust = 1))\n",
    "pdf(paste0(figures.dir, \"6H Treg CTLA4 gene expression.pdf\"), width = 10, height = 5) \n",
    "print(p)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3808d424-9dc0-49cf-abb7-e0f195327d50",
   "metadata": {},
   "source": [
    "# Plot exhaustion/activation signatures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "b4ef6c0f-ddb3-4749-af77-c340e8fe8d60",
   "metadata": {},
   "outputs": [],
   "source": [
    "csf.high.dose$Patient <- factor(csf.high.dose$Patient, levels = all.high.dose.patients)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "f4457d3c-63bf-42e5-9cc2-e85b497e05d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "wherry_signatures <- read.csv(paste0(root.dir, \"Resources/\", \"wherry_pace_signatures.csv\"))\n",
    "effector.signature <- wherry_signatures %>% pull(Pace_effector_markers)\n",
    "effector.signature <- effector.signature[effector.signature != \"\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "14d74955-df42-4e19-91e8-1089e2238618",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“\u001b[1m\u001b[22mThe `slot` argument of `GetAssayData()` is deprecated as of SeuratObject 5.0.0.\n",
      "\u001b[36mℹ\u001b[39m Please use the `layer` argument instead.\n",
      "\u001b[36mℹ\u001b[39m The deprecated feature was likely used in the \u001b[34mSeurat\u001b[39m package.\n",
      "  Please report the issue at \u001b[3m\u001b[34m<https://github.com/satijalab/seurat/issues>\u001b[39m\u001b[23m.”\n",
      "Warning message:\n",
      "“The following features are not present in the object: CCL9, KLRA9, ATPIF1, attempting to find updated synonyms”\n",
      "Found updated symbols for 1 symbols\n",
      "\n",
      "ATPIF1 -> ATP5IF1\n",
      "\n",
      "Warning message:\n",
      "“The following features are still not present in the object: CCL9, KLRA9”\n"
     ]
    }
   ],
   "source": [
    "gene.signatures <- list()\n",
    "gene.signatures[[\"Effector signature\"]] <- effector.signature\n",
    "csf.high.dose <- AddModuleScore(csf.high.dose, feature=gene.signatures, name=names(gene.signatures), search = T)\n",
    "csf.high.dose@meta.data <- csf.high.dose@meta.data %>% rename(`Effector signature`=`Effector signature1`)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "b5e8d4e1-ae66-4295-841f-9ec8bf85c8f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "data <- FetchData(csf.high.dose, vars = c(\"HLA-DRA\", \"ENTPD1\", \"CTLA4\", \"PDCD1\", \"LAG3\"))\n",
    "csf.high.dose <- AddMetaData(csf.high.dose, data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "ca42cf74-7e4d-43a4-99bd-531902de8f31",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'cell_type', 'sample_id', 'Patient',\n",
      "'response.status'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'cell_type', 'sample_id', 'Patient',\n",
      "'response.status'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'cell_type', 'sample_id', 'Patient',\n",
      "'response.status'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'cell_type', 'sample_id', 'Patient',\n",
      "'response.status'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'cell_type', 'sample_id', 'Patient',\n",
      "'response.status'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'cell_type', 'sample_id', 'Patient',\n",
      "'response.status'. You can override using the `.groups` argument.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Figure 6J\n",
    "genes.of.interest <- c(\"HLA-DRA\", \"ENTPD1\", \"CTLA4\",  \"Effector signature\", \"PDCD1\", \"LAG3\")\n",
    "\n",
    "plot.list <- list()\n",
    "celltype <- \"CD8+ T\"\n",
    "day <- \"D7\"\n",
    "timepoint.var <- \"Day\"\n",
    "seurat.obj <- csf.high.dose\n",
    "\n",
    "for (gene.of.interest in genes.of.interest) {\n",
    "\n",
    "    seurat.obj$gene.of.interest <- seurat.obj[[gene.of.interest]]\n",
    "    gene.df <- seurat.obj@meta.data %>% \n",
    "        filter(cell_type == celltype, .data[[timepoint.var]] == day) %>%\n",
    "        group_by(cell_type, sample_id, Patient, response.status, Cohort) %>% \n",
    "        summarize(mean.expression = mean(gene.of.interest),\n",
    "                  count = n()) %>%\n",
    "        ungroup()\n",
    "    \n",
    "    stat.test <- gene.df %>%\n",
    "      wilcox_test(mean.expression ~ response.status) %>%\n",
    "      adjust_pvalue(method = \"fdr\") %>%\n",
    "      add_significance(\"p.adj\") %>%\n",
    "      add_y_position()\n",
    "    \n",
    "    p <- ggplot(gene.df, aes(response.status, mean.expression)) +\n",
    "        geom_boxplot(aes(fill = response.status), outlier.shape = NA) + \n",
    "        geom_point(aes(shape = Cohort), stroke = 1, size = 3, fill = \"white\", color = \"black\", position = position_jitter(width=0.2)) + \n",
    "        scale_shape_manual(breaks = c(\"Discovery\", \"Validation\"), values = c(19, 22)) + \n",
    "        scale_fill_manual(breaks = c(\"Non-responder\", \"Responder\"), values = c(\"#A0412D\", \"#5FA7A2\")) +\n",
    "        theme_classic() +\n",
    "        ylab(\"Expression level\") +\n",
    "        ggtitle(gene.of.interest) +\n",
    "        stat_pvalue_manual(stat.test, label = \"p.adj\", tip.length = 0) +\n",
    "        xlab(\"Response status\") +\n",
    "        guides(fill = \"none\")\n",
    "    \n",
    "    plot.list[[gene.of.interest]] <- p\n",
    "}\n",
    "\n",
    "p <- plot_grid(plotlist = plot.list)\n",
    "pdf(paste0(figures.dir, \"6J CD8+ T Day 7 marker expression.pdf\"), width=12, height=9)\n",
    "print(p)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "85c95f91-5248-494b-ae8c-2018f74668f7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.3.2 (2023-10-31)\n",
       "Platform: aarch64-unknown-linux-gnu (64-bit)\n",
       "Running under: Ubuntu 22.04.3 LTS\n",
       "\n",
       "Matrix products: default\n",
       "BLAS:   /usr/lib/aarch64-linux-gnu/openblas-pthread/libblas.so.3 \n",
       "LAPACK: /usr/lib/aarch64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "time zone: Etc/UTC\n",
       "tzcode source: system (glibc)\n",
       "\n",
       "attached base packages:\n",
       "[1] stats     graphics  grDevices utils     datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] rstatix_0.7.2      cowplot_1.1.3      ggpubr_0.6.0       ggplot2_3.5.2     \n",
       " [5] tidyr_1.3.1        dplyr_1.1.4        liana_0.1.14       Seurat_5.3.0      \n",
       " [9] SeuratObject_5.1.0 sp_2.2-0          \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] fs_1.6.6                    matrixStats_1.5.0          \n",
       "  [3] spatstat.sparse_3.1-0       bitops_1.0-9               \n",
       "  [5] lubridate_1.9.4             httr_1.4.7                 \n",
       "  [7] RColorBrewer_1.1-3          doParallel_1.0.17          \n",
       "  [9] repr_1.1.7                  tools_4.3.2                \n",
       " [11] sctransform_0.4.2           backports_1.5.0            \n",
       " [13] R6_2.6.1                    lazyeval_0.2.2             \n",
       " [15] uwot_0.2.3                  GetoptLong_1.0.5           \n",
       " [17] withr_3.0.2                 prettyunits_1.2.0          \n",
       " [19] gridExtra_2.3               progressr_0.15.1           \n",
       " [21] cli_3.6.5                   Biobase_2.62.0             \n",
       " [23] spatstat.explore_3.4-2      fastDummies_1.7.5          \n",
       " [25] labeling_0.4.3              spatstat.data_3.1-6        \n",
       " [27] readr_2.1.5                 ggridges_0.5.6             \n",
       " [29] pbapply_1.7-2               pbdZMQ_0.3-14              \n",
       " [31] R.utils_2.13.0              dichromat_2.0-0.1          \n",
       " [33] parallelly_1.43.0           sessioninfo_1.2.3          \n",
       " [35] limma_3.58.1                readxl_1.4.5               \n",
       " [37] RSQLite_2.3.11              generics_0.1.3             \n",
       " [39] shape_1.4.6.1               ica_1.0-3                  \n",
       " [41] spatstat.random_3.3-3       car_3.1-3                  \n",
       " [43] zip_2.3.2                   Matrix_1.6-4               \n",
       " [45] S4Vectors_0.40.2            logger_0.4.0               \n",
       " [47] abind_1.4-8                 R.methodsS3_1.8.2          \n",
       " [49] lifecycle_1.0.4             edgeR_4.0.3                \n",
       " [51] yaml_2.3.10                 carData_3.0-5              \n",
       " [53] SummarizedExperiment_1.32.0 SparseArray_1.2.4          \n",
       " [55] Rtsne_0.17                  grid_4.3.2                 \n",
       " [57] blob_1.2.4                  dqrng_0.4.1                \n",
       " [59] promises_1.3.2              crayon_1.5.3               \n",
       " [61] dir.expiry_1.8.0            miniUI_0.1.2               \n",
       " [63] lattice_0.21-9              beachmat_2.18.1            \n",
       " [65] metapod_1.8.0               pillar_1.10.2              \n",
       " [67] knitr_1.50                  ComplexHeatmap_2.18.0      \n",
       " [69] GenomicRanges_1.54.1        tcltk_4.3.2                \n",
       " [71] rjson_0.2.23                future.apply_1.11.3        \n",
       " [73] codetools_0.2-19            glue_1.8.0                 \n",
       " [75] spatstat.univar_3.1-2       data.table_1.17.0          \n",
       " [77] vctrs_0.6.5                 png_0.1-8                  \n",
       " [79] spam_2.11-1                 cellranger_1.1.0           \n",
       " [81] gtable_0.3.6                cachem_1.1.0               \n",
       " [83] OmnipathR_3.15.14           xfun_0.52                  \n",
       " [85] S4Arrays_1.2.1              mime_0.13                  \n",
       " [87] survival_3.5-7              SingleCellExperiment_1.24.0\n",
       " [89] iterators_1.0.14            statmod_1.5.0              \n",
       " [91] bluster_1.10.0              fitdistrplus_1.2-2         \n",
       " [93] ROCR_1.0-11                 nlme_3.1-163               \n",
       " [95] bit64_4.6.0-1               progress_1.2.3             \n",
       " [97] filelock_1.0.3              RcppAnnoy_0.0.22           \n",
       " [99] GenomeInfoDb_1.38.8         irlba_2.3.5.1              \n",
       "[101] KernSmooth_2.23-22          colorspace_2.1-1           \n",
       "[103] BiocGenerics_0.48.1         DBI_1.2.3                  \n",
       "[105] tidyselect_1.2.1            bit_4.6.0                  \n",
       "[107] compiler_4.3.2              curl_6.2.2                 \n",
       "[109] rvest_1.0.4                 httr2_1.1.2                \n",
       "[111] BiocNeighbors_1.20.2        basilisk.utils_1.21.0      \n",
       "[113] xml2_1.3.8                  DelayedArray_0.28.0        \n",
       "[115] plotly_4.10.4               checkmate_2.3.2            \n",
       "[117] scales_1.4.0                lmtest_0.9-40              \n",
       "[119] rappdirs_0.3.3              stringr_1.5.1              \n",
       "[121] digest_0.6.37               goftest_1.2-3              \n",
       "[123] spatstat.utils_3.1-3        rmarkdown_2.29             \n",
       "[125] basilisk_1.21.0             XVector_0.42.0             \n",
       "[127] htmltools_0.5.8.1           pkgconfig_2.0.3            \n",
       "[129] base64enc_0.1-3             sparseMatrixStats_1.14.0   \n",
       "[131] MatrixGenerics_1.14.0       fastmap_1.2.0              \n",
       "[133] rlang_1.1.6                 GlobalOptions_0.1.2        \n",
       "[135] htmlwidgets_1.6.4           DelayedMatrixStats_1.24.0  \n",
       "[137] shiny_1.10.0                farver_2.1.2               \n",
       "[139] zoo_1.8-14                  jsonlite_2.0.0             \n",
       "[141] BiocParallel_1.36.0         R.oo_1.27.1                \n",
       "[143] BiocSingular_1.18.0         RCurl_1.98-1.17            \n",
       "[145] magrittr_2.0.3              Formula_1.2-5              \n",
       "[147] scuttle_1.12.0              GenomeInfoDbData_1.2.11    \n",
       "[149] dotCall64_1.2               patchwork_1.3.0            \n",
       "[151] IRkernel_1.3.2              Rcpp_1.0.14                \n",
       "[153] reticulate_1.42.0           stringi_1.8.7              \n",
       "[155] zlibbioc_1.48.2             MASS_7.3-60                \n",
       "[157] plyr_1.8.9                  parallel_4.3.2             \n",
       "[159] listenv_0.9.1               ggrepel_0.9.6              \n",
       "[161] deldir_2.0-4                IRdisplay_1.1              \n",
       "[163] splines_4.3.2               tensor_1.5                 \n",
       "[165] hms_1.1.3                   circlize_0.4.16            \n",
       "[167] locfit_1.5-9.12             igraph_2.1.4               \n",
       "[169] uuid_1.2-1                  spatstat.geom_3.3-6        \n",
       "[171] ggsignif_0.6.4              RcppHNSW_0.6.0             \n",
       "[173] reshape2_1.4.4              stats4_4.3.2               \n",
       "[175] ScaledMatrix_1.10.0         XML_3.99-0.18              \n",
       "[177] evaluate_1.0.3              scran_1.28.2               \n",
       "[179] tzdb_0.5.0                  foreach_1.5.2              \n",
       "[181] httpuv_1.6.16               RANN_2.6.2                 \n",
       "[183] purrr_1.0.4                 polyclip_1.10-7            \n",
       "[185] future_1.40.0               clue_0.3-66                \n",
       "[187] scattermore_1.2             rsvd_1.0.5                 \n",
       "[189] broom_1.0.8                 xtable_1.8-4               \n",
       "[191] RSpectra_0.16-2             later_1.4.2                \n",
       "[193] viridisLite_0.4.2           tibble_3.2.1               \n",
       "[195] memoise_2.0.1               IRanges_2.36.0             \n",
       "[197] cluster_2.1.6               timechange_0.3.0           \n",
       "[199] globals_0.17.0             "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2938289-d935-48d5-944f-9e9c60b7e814",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
